processing/src/main/java/io/druid/segment/VirtualColumns.java

```diff
@@ -19,62 +19,250 @@
+    }
+  }
+
+  public FloatColumnSelector makeFloatColumnSelector(String columnName, ColumnSelectorFactory factory)
+  {
+    final VirtualColumn virtualColumn = getVirtualColumn(columnName);
+    if (virtualColumn == null) {
+      return ZeroFloatColumnSelector.instance();
+    } else {
+      final FloatColumnSelector selector = virtualColumn.makeFloatColumnSelector(columnName, factory);
+      return selector == null ? ZeroFloatColumnSelector.instance() : selector;
+    }
+  }
+
+  public LongColumnSelector makeLongColumnSelector(String columnName, ColumnSelectorFactory factory)
+  {
+    final VirtualColumn virtualColumn = getVirtualColumn(columnName);
+    if (virtualColumn == null) {
+      return ZeroLongColumnSelector.instance();
+    } else {
+      final LongColumnSelector selector = virtualColumn.makeLongColumnSelector(columnName, factory);
+      return selector == null ? ZeroLongColumnSelector.instance() : selector;
     }
-    return withDotSupport.get(dimension);
+  }
+
+  public ColumnCapabilities getColumnCapabilities(String columnName)
+  {
+    final VirtualColumn virtualColumn = getVirtualColumn(columnName);
+    return virtualColumn != null
+           ? Preconditions.checkNotNull(virtualColumn.capabilities(columnName), "capabilities for column[%s]", columnName)
+           : null;
   }
 
   public boolean isEmpty()
   {
     return withDotSupport.isEmpty() && withoutDotSupport.isEmpty();
   }
+
+  @JsonValue
+  public VirtualColumn[] getVirtualColumns()
+  {
+    // VirtualColumn[] instead of List<VirtualColumn> to aid Jackson serialization.
+    return virtualColumns.toArray(new VirtualColumn[]{});
+  }
+
+  public ColumnSelectorFactory wrap(final ColumnSelectorFactory baseFactory)
+  {
+    return new VirtualizedColumnSelectorFactory(baseFactory, this);
+  }
+
+  public byte[] getCacheKey()
+  {
+    final byte[][] cacheKeys = new byte[virtualColumns.size()][];
+    int len = Ints.BYTES;
+    for (int i = 0; i < virtualColumns.size(); i++) {
+      cacheKeys[i] = virtualColumns.get(i).getCacheKey();
+      len += Ints.BYTES + cacheKeys[i].length;
+    }
+    final ByteBuffer buf = ByteBuffer.allocate(len).putInt(virtualColumns.size());
+    for (byte[] cacheKey : cacheKeys) {
+      buf.putInt(cacheKey.length);
+      buf.put(cacheKey);
+    }
+    return buf.array();
+  }
+
+  private void detectCycles(VirtualColumn virtualColumn, ImmutableSet<String> columnNames)
+  {
+    Set<String> nextSet = columnNames == null
+                          ? Sets.newHashSet(virtualColumn.getOutputName())
+                          : Sets.newHashSet(columnNames);
+
+    for (String columnName : virtualColumn.requiredColumns()) {
+      if (!nextSet.add(columnName)) {
+        throw new IAE("Self-referential column[%s]", columnName);
+      }
+
+      final VirtualColumn dependency = getVirtualColumn(columnName);
+      if (dependency != null) {
+        detectCycles(dependency, ImmutableSet.copyOf(nextSet));
```

Comment:
1. It's added above on the line `if (!nextSet.add(columnName)) {`
2. Reworked to avoid copying